### PR TITLE
Expose job enqueueing / running API

### PR DESCRIPTION
**Pull request by jprendes**
Expose QuickJS's API to enqueue and run jobs in the context.

The job running API is part of the `JSRuntime`, and returns as output the JSContext that the job is part of.
Enabling multi-context workflows requires adding context information to `exception`.

### CODE REVIEW

File: quickjspp.hpp [hunk 1]
```diff
@@ -951,7 +957,13 @@ struct js_traits<T *, std::enable_if_t<std::is_class_v<T>>>
...
         {
 #if defined(__cpp_rtti)
-            js_traits<std::shared_ptr<T>>::register_class(ctx, typeid(T).name());
+            if constexpr(std::is_same_v<std::decay_t<T> *, JSValue>) {
+                // This needs to be a special case because in `CONFIG_CHECK_JSVALUE` mode
+                // `JSValue` is a pointer to an incomplete type, which doesn't support `typeid`.
+                js_traits<std::shared_ptr<T>>::register_class(ctx, "JSValue");
+            } else {
+                js_traits<std::shared_ptr<T>>::register_class(ctx, typeid(T).name());
+            }
```
Comment (ftk) [inline quickjspp.hpp:966]:
```suggestion
            // If you have an error here with T=JSValueConst
            // it probably means you are passing JSValueConst to where JSValue is expected
            js_traits<std::shared_ptr<T>>::register_class(ctx, typeid(T).name());
```

File: quickjspp.hpp [hunk 2]
```diff
@@ -1868,4 +1888,48 @@ property_proxy<Key>::operator Value() const
...
 
+inline void Context::enqueueJob(std::function<void()> && job_fcn) {
+    // No need to free the new value. It will be freed automatically when
+    // the function is called.
+    JSValueConst job_val = newValue(std::move(job_fcn)).release();
+    JS_EnqueueJob(ctx, [](JSContext *ctx, int argc, JSValueConst *argv){
+        try
+        {
+            // use Value constructor rather than context.newValue to avoid retrieving the Context object.
+            Value{ctx, argv[0]}.as<std::function<void()>>()();
```
Comment (ftk) [inline quickjspp.hpp:1899]: Correct code would be
`Value{ctx, JS_DupValue(ctx, argv[0])}.as<std::function<void()>>()();` which also won't free argv[0]

File: quickjspp.hpp [hunk 3]
```diff
@@ -1868,4 +1888,48 @@ property_proxy<Key>::operator Value() const
...
+            JS_ThrowInternalError(ctx, "%s", err.what());
+            return JS_EXCEPTION;
+        }
+        catch (...)
+        {
+            JS_ThrowInternalError(ctx, "Unknown error");
+            return JS_EXCEPTION;
+        }
+        return JS_UNDEFINED;
+    }, 1, &job_val);
```
Comment (ftk) [inline quickjspp.hpp:1916]:
```suggestion
    auto job_val = newValue(std::move(job_fcn));
    JSValueConst arg = job_val.v;
    int err = JS_EnqueueJob(ctx, [](JSContext *ctx, int argc, JSValueConst *argv){
        try
        {
            assert(argc >= 1);
            js_traits<std::function<void()>>::unwrap(ctx, argv[0])();
        }
        catch (exception)
        {
            return JS_EXCEPTION;
        }
        catch (std::exception const & err)
        {
            JS_ThrowInternalError(ctx, "%s", err.what());
            return JS_EXCEPTION;
        }
        catch (...)
        {
            JS_ThrowInternalError(ctx, "Unknown error");
            return JS_EXCEPTION;
        }
        return JS_UNDEFINED;
    }, 1, &arg);
    if(err < 0)
        throw exception{ctx};
```
job_fcn can also be optimized to not be wrapped in std::function but whatever.

File: quickjspp.hpp [hunk 4]
```diff
@@ -1356,6 +1368,12 @@ class Runtime
     {
         JS_FreeRuntime(rt);
     }
+
+    Context * executePendingJob();
```
Comment (ftk) [inline quickjspp.hpp:1372]:
```suggestion

    /// @return pointer to qjs::Context of the executed job or nullptr if no job is pending
    Context * executePendingJob();
```

**Review by ftk — APPROVED** [review]

File: quickjspp.hpp (hunk 3 above)
Comment (jprendes) [reply quickjspp.hpp:1916]: you mean using a templated `Functor`?

File: quickjspp.hpp (hunk 2 above)
Comment (jprendes) [reply quickjspp.hpp:1900]: Ahhh, I see. Thanks!

File: quickjspp.hpp (hunk 3 above)
Comment (ftk) [reply quickjspp.hpp:1917]: Yes, something like this
```cpp
    template <typename Function>
    void enqueueJob(Function&& f)
    {
         JSValue job_val = js_traits<std::function<void()>>::wrap(ctx, std::forward<Function>(f));
...
JS_FreeValue(job_val);
```

File: test/jobs.cpp [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+    try {
+        while (runtime.isJobPending()) {
+            runtime.executePendingJob();
+        }
+        assert(false && "Job execution should have failed");
+    } catch (qjs::exception const & exc) {
+        assert(&exc.context() == &context && "Exception should contain context information");
+    }
+
```
Comment (ftk) [inline test/jobs.cpp:43]:
```suggestion
    // Check that not executed jobs won't introduce memory leaks
    context.enqueueJob([](){
        assert(false && "This job will not be called");
    });
```

**Comment by jprendes** [comment]
I think all the suggestions have been taken into account now.
GitHub decided to act up yesterday evening, I wasn't able to push any changes!

**Pull request merged**